pkg/comp-functions/functions/vshnredis/alerting.go

```diff
@@ -0,0 +1,140 @@
+package vshnredis
+
+import (
+	"context"
+
+	controllerruntime "sigs.k8s.io/controller-runtime"
+
+	xkube "github.com/crossplane-contrib/provider-kubernetes/apis/object/v1alpha1"
+	alertmanagerv1alpha1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1alpha1"
+	vshnv1 "github.com/vshn/appcat/v4/apis/vshn/v1"
+	"github.com/vshn/appcat/v4/pkg/comp-functions/runtime"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+)
+
+// AddUserAlerting adds user alerting to the Redis instance.
+func AddUserAlerting(ctx context.Context, iof *runtime.Runtime) runtime.Result {
+	log := controllerruntime.LoggerFrom(ctx)
+	log.Info("Check if alerting references are set")
+
+	log.V(1).Info("Transforming", "obj", iof)
+
+	err := runtime.AddToScheme(alertmanagerv1alpha1.SchemeBuilder)
+	if err != nil {
+		return runtime.NewFatalErr(ctx, "Cannot add scheme builder to scheme", err)
+	}
+	comp := &vshnv1.VSHNRedis{}
+	err = iof.Observed.GetComposite(ctx, comp)
+	if err != nil {
+		return runtime.NewFatalErr(ctx, "Cannot get composite from function io", err)
+	}
+
+	// Wait for the next reconciliation in case instance namespace is missing
+	if comp.Status.InstanceNamespace == "" {
+		return runtime.NewWarning(ctx, "Composite is missing instance namespace, skipping transformation")
+	}
+
+	monitoringSpec := comp.Spec.Parameters.Monitoring
+
+	if monitoringSpec.AlertmanagerConfigRef != "" {
+
+		if monitoringSpec.AlertmanagerConfigSecretRef == "" {
+			return runtime.NewFatal(ctx, "Found AlertmanagerConfigRef but no AlertmanagerConfigSecretRef, please specify as well")
+		}
+
+		refName := comp.Spec.Parameters.Monitoring.AlertmanagerConfigRef
+		log.Info("Found an AlertmanagerConfigRef, deploying...", "refName", refName)
+
+		err = deployAlertmanagerFromRef(ctx, comp, iof)
+		if err != nil {
+			return runtime.NewFatalErr(ctx, "Could not deploy alertmanager from ref", err)
+		}
+	}
+
+	if monitoringSpec.AlertmanagerConfigSpecTemplate != nil {
+
+		if monitoringSpec.AlertmanagerConfigSecretRef == "" {
+			return runtime.NewFatal(ctx, "Found AlertmanagerConfigTemplate but no AlertmanagerConfigSecretRef, please specify as well")
+		}
+
+		log.Info("Found an AlertmanagerConfigTemplate, deploying...")
+
+		err = deployAlertmanagerFromTemplate(ctx, comp, iof)
+		if err != nil {
+			return runtime.NewFatalErr(ctx, "Cannot deploy alertmanager from template", err)
+		}
+	}
+
+	if monitoringSpec.AlertmanagerConfigSecretRef != "" {
+		refName := comp.Spec.Parameters.Monitoring.AlertmanagerConfigSecretRef
+		log.Info("Found an AlertmanagerConfigSecretRef, deploying...", "refName", refName)
+
+		err = deploySecretRef(ctx, comp, iof)
+		if err != nil {
+			return runtime.NewFatalErr(ctx, "Cannot deploy secret ref", err)
+		}
+	}
+
+	return runtime.NewNormal()
+}
+
+func deployAlertmanagerFromRef(ctx context.Context, comp *vshnv1.VSHNRedis, iof *runtime.Runtime) error {
```

Comment:
Can't we work with generics here and put this function in a common place and reuse it for both services?
It's basically still a copy&paste from the postgreSQL FunctionIO